### ISSUE TITLE
Torch and Lightning 2.0.0

### ISSUE THREAD

**Issue by kylebgorman** [CUNY-CL/yoyodyne#60]
The library does not work with Lightning (and one suspects that Torch itself is also an issue) > 2.0.0. The first issue I encounter when running `yoyodyne-train` with no arguments is related to a change in how Lightning command-line arguments are handled---I suspect there are at least a few more.

So that the library is not broken at head---which I consider unacceptable---I have pinned as follows:

```
pytorch-lightning>=1.7.0,<2.0.0
torch>=1.11.0,<2.0.0
```

What we need to do is just to migrate to 2.0.0, by fixing Lightning (and Torch, if any) bugs until things work, and then re-pin these two dependencies `>=2.0.0`. I have initially assigned myself, but I would welcome help.

**Comment by kylebgorman** [comment]
I read these docs with some care. There are two steps:

1. Before migrating to 2.0, we should should implement a subclass of `LightningDataModule` for training:

**Comment by kylebgorman** [comment]
I have read these docs and have the basic lay of the land now. The migration should go in two steps:

1. First, we should implement subclasses of `LightningDataModule` (this is available in Lightning pre 2.0 too). The `train_dataloader` method should return the training data dataloader (naturally) and similarly with `dev_dataloader` and `test_dataloader`. (Even though we don't need access to train, dev, and test in any one process, we can probably set it up so we can use the same one for training and inference.) Most of the logic of [`get_datasets`](https://github.com/CUNY-CL/yoyodyne/blob/035f2224c9ef075e1c909b20e4a08face509c1aa/yoyodyne/train.py#L180C1-L181C1) and [`get_loaders`](https://github.com/CUNY-CL/yoyodyne/blob/035f2224c9ef075e1c909b20e4a08face509c1aa/yoyodyne/train.py#L181) would be moved into this too. This is thus mostly just cleanup.
2. Then, we have to move arg parsing to `LightningCli` and migrate Torch and Lightning to >=2.0. This is a bit magical (it introspects to find flags for the dataset and models) but our house is mostly in order so it might just work. At this point there are probably one or two low-level Torch issues we'll hit up against too and can handle those as they arise...

**Comment by Adamits** [comment]
Sounds good, I think I understand. Though I feel a bit annoyed that we have to use `LightningCli` instead of standard python libraries... I will try to prioritize it this week.

**Comment by kylebgorman** [comment]
> Sounds good, I think I understand. Though I feel a bit annoyed that we have to use `LightningCli` instead of standard python libraries... I will try to prioritize it this week.

I'm working on the data module part of things in the next few days---haven't thought about the model side yet.

FWIW, what LightningCLI is is basically a tool that can introspect to populate the argparse parser...which should help us enormously with the kind of bug we saw with not passing around the label smoothing...

**Comment by kylebgorman** [comment]
@Adamits believes it is possible to migrate to Torch 2 while staying on the 1.13 branch of PTL. The ability to complete this migration in two steps (one to Torch 2, then later migrating to PTL 2 focusing on CLI changes) would make this much easier.

**Comment by kylebgorman** [comment]
Draft of the migration to PyTorch 2 in #173. Testing now---focusing on the transformer since we can use the causal mask flag there, hopefully.

**Comment by kylebgorman** [comment]
I have been using LightningCLI in another project and I have to say, it's a total vibe shift and it's way better than having big Bash scripts (without any nice syntax support), so I think we should prioritize this now.

**Comment by kylebgorman** [comment]
Another feature of this migration is that at least some if not all the models will work on M1/M2 Macs using their MPS accelerators.
